### ISSUE TITLE
Putting common fields inside type unions

### ISSUE THREAD

**Issue by ktosiek** [relay-tools/relay-compiler-language-typescript#100]
Currently asking for different fragments and __typename creates a union of types for each fragment. But this mechanism disengages when I select new fields:

```
fragment ConcreateTypes on Viewer {
  actor {
    __typename
    id
    ... on Page {
      ...PageFragment
    }
    ... on User {
      name
    }
  }
}
```

Current result is:
```
export type ConcreateTypes = {
    readonly actor: ({
        readonly __typename: string;
        readonly id: string;
        readonly name?: string | null;
        readonly " $fragmentRefs": PageFragment$ref;
    }) | null;
    readonly " $refType": ConcreateTypes$ref;
};
```

I would expect the result look like:
```
export type ConcreateTypes = {
    readonly actor: ({
        readonly __typename: "Page";
        readonly id: string;
        readonly " $fragmentRefs": PageFragment$ref;
    } | {
        readonly __typename: "User";
        readonly id: string;
        readonly name: string | null;
    } | {
        /*This will never be '% other', but we need some
        value in case none of the concrete values match.*/
        readonly __typename: "%other";
        readonly id: string;
    }) | null;
};
```

**Comment by ktosiek** [comment]
Looks like this is related to https://github.com/facebook/relay/issues/2039

**Comment by ktosiek** [comment]
Would it make sense to have both the union and flattened types, intersected?
Then one could either use the fields as optional, or check the typename and get better guarantees.

**Comment by alloy** [comment]
Fixed in #103.
